File: yumi/sdk/rust/src/schema.rs
use crate::types::RegisterOptions;
use serde_json::{json, Value};

pub fn build_tool_schema(opts: &RegisterOptions) -> Value {
    let mut properties = serde_json::Map::new();
    let mut required = Vec::new();

    for p in &opts.parameters {
        properties.insert(
            p.name.clone(),
            json!({
                "type": p.type_name,
                "description": p.description,
            }),
        );
        let is_required = p.required.unwrap_or(true);
        if is_required {
            required.push(p.name.clone());
        }
    }

    let mut schema = json!({
        "type": "function",
        "function": {
            "name": opts.name,
            "description": opts.description,
            "parameters": {
                "type": "object",
                "properties": properties,
                "required": required,
            }
        }
    });

    if let Some(t) = opts.timeout {
        schema
            .as_object_mut()
            .unwrap()
            .insert("timeout".to_string(), json!(t));
    }
    if opts.require_confirmation {
        schema
            .as_object_mut()
            .unwrap()
            .insert("require_confirmation".to_string(), json!(true));
    }
    if opts.always_include {
        schema
            .as_object_mut()
            .unwrap()
            .insert("always_include".to_string(), json!(true));
    }
    if opts.allow_proactive {
        schema
            .as_object_mut()
            .unwrap()
            .insert("allow_proactive".to_string(), json!(true));
    }
    if opts.proactive_context {
        schema
            .as_object_mut()
            .unwrap()
            .insert("proactive_context".to_string(), json!(true));
    }
    if let Some(args) = &opts.proactive_context_args {
        schema
            .as_object_mut()
            .unwrap()
            .insert("proactive_context_args".to_string(), args.clone());
    }
    if let Some(description) = &opts.proactive_context_description {
        schema
            .as_object_mut()
            .unwrap()
            .insert("proactive_context_description".to_string(), json!(description));
    }

    schema
}

File: yumi/sdk/dart/lib/src/schema.dart
import 'types.dart';

Map<String, dynamic> buildToolSchema(RegisterOptions opts) {
  // Map the `mode` API onto the existing wire flags (one mode per tool).
  var alwaysInclude = opts.alwaysInclude;
  var proactiveContext = opts.proactiveContext;
  var proactiveContextArgs = opts.proactiveContextArgs;
  var proactiveContextDescription = opts.proactiveContextDescription;
  switch (opts.mode) {
    case 'dynamic':
      break;
    case 'pinned':
      alwaysInclude = true;
      break;
    case 'autorun':
      proactiveContext = true;
      if (opts.contextArgs != null) proactiveContextArgs = opts.contextArgs;
      if (opts.contextLabel != null) {
        proactiveContextDescription = opts.contextLabel;
      }
      break;
    default:
      throw ArgumentError(
        "mode must be 'dynamic', 'pinned', or 'autorun'; got '${opts.mode}'",
      );
  }

  final properties = <String, dynamic>{};
  final required = <String>[];
  for (final p in opts.parameters) {
    properties[p.name] = {
      'type': p.typeName,
      'description': p.description,
    };
    final isRequired = p.required_ ?? true;
    if (isRequired) required.add(p.name);
  }

  final schema = <String, dynamic>{
    'type': 'function',
    'function': {
      'name': opts.name,
      'description': opts.description,
      'parameters': {
        'type': 'object',
        'properties': properties,
        'required': required,
      },
    },
  };
  if (opts.timeout != null) {
    schema['timeout'] = opts.timeout;
  }
  if (opts.requireConfirmation) {
    schema['require_confirmation'] = true;
  }
  if (alwaysInclude) {
    schema['always_include'] = true;
  }
  if (opts.allowProactive) {
    schema['allow_proactive'] = true;
  }
  if (proactiveContext) {
    schema['proactive_context'] = true;
  }
  if (proactiveContextArgs != null) {
    schema['proactive_context_args'] = proactiveContextArgs;
  }
  if (proactiveContextDescription != null &&
      proactiveContextDescription.trim().isNotEmpty) {
    schema['proactive_context_description'] = proactiveContextDescription;
  }
  return schema;
}
